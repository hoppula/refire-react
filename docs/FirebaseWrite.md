# FirebaseWrite

> Higher order component that enables easy write operations to Firebase

## Options

**path**

Firebase root path for write operation

**method**

Firebase write method, one of `push`, `set`, `transaction` or `update`.


```js
import { FirebaseWrite } from 'refire-react'
class CommentComponent extends Component {
  render() {
    const {
      submit,
      clearErrors,
      errors,
      processing
    } = this.props
  }
}
export default FirebaseWrite({
  path: "",
  method: "update"
})(CommentComponent)
```
