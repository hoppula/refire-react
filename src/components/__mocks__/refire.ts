import {Â firebaseActions as actions } from "refire"
export { syncFirebase, firebaseToProps, firebaseReducer } from "refire"
export const firebaseActions = {
  ...actions,
  passwordLogin: jest.fn((email, password) => {
    return (dispatch, getState) => {
      return Promise.resolve()
    }
  }),
  createUser: jest.fn((email, password) => {
    return (dispatch, getState) => {
      return Promise.resolve()
    }
  }),
  resetPassword: jest.fn((email) => {
    return (dispatch, getState) => {
      return Promise.resolve()
    }
  }),
  write: jest.fn((method, path, value, ownProps) => {
    return (dispatch, getState) => {
      return Promise.resolve()
    }
  }),
  clearWriteErrors: jest.fn((path) => {
    return (dispatch, getState) => {
      return Promise.resolve()
    }
  }),
}
