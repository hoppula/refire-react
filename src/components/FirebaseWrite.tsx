import * as React from "react"
import { Component } from "react"
import * as ReactRedux from "react-redux"
import { connect } from "react-redux"
import { firebaseActions } from "refire"

const { clearWriteErrors, write } = firebaseActions

interface ValidMethods {
  [key: string]: boolean
}

const validMethods: ValidMethods = {
  push: true,
  set: true,
  transaction: true,
  update: true
}

export interface WriteProps {
  readonly dispatch?: ReactRedux.Dispatch<{}>,
  readonly processing?: string[],
  readonly errors?: string[],
  readonly path: string
}

export interface FirebaseWriteState {
  firebase: {
    writes: {
      processing: {
        [name: string]: string[]
      },
      errors: {
        [name: string]: string[]
      }
    }
  }
}

export type pathFunction = (state: FirebaseWriteState, ownProps: any) => string

export interface WriteOptions {
  path?: string | pathFunction,
  method?: "" | "push" | "set" | "transaction" | "update"
}

export interface ExtraProps {
  clearErrors?: () => void,
  errors?: string[],
  processing?: boolean,
  submit?: (value: any) => void
}

export default function(options: WriteOptions = {}) {

  const { path = "", method } = options

  if (typeof path !== "function" && typeof path !== "string") {
    throw new Error("options.path must be a function or string")
  }

  if (typeof method !== "string" || !validMethods[method]) {
    throw new Error(`options.method must be one of: ${Object.keys(validMethods).join(", ")}`)
  }

  function mapStateToProps(state: FirebaseWriteState, ownProps: any) {
    const processing = state.firebase.writes.processing
    const errors = state.firebase.writes.errors
    const firebasePath: string = typeof path === "function"
      ? path(state, ownProps)
      : path

    return {
      errors: errors[firebasePath],
      path: firebasePath,
      processing: processing[firebasePath]
    }
  }

  return function <P>(
    WrappedComponent: ReactRedux.Component<P>
  ) {

    class FirebaseWrite extends Component<WriteProps & P, {}> {

      submit = (value: any) => {
        const { dispatch } = this.props

        return dispatch(
          write({ method, path, value, ownProps: this.props })
        )
      }

      clearErrors = () => {
        this.props.dispatch(
          clearWriteErrors(this.props.path)
        )
      }

      render() {
        const processing = this.props.processing && !!this.props.processing.length
        const errors = this.props.errors || []

        const extraProps: ExtraProps = {
          clearErrors: this.clearErrors,
          errors,
          processing,
          submit: this.submit
        }

        return (
          <WrappedComponent
            { ...this.props }
            { ...extraProps }
          />
        )
      }

    }

    return connect<WriteProps, void, P>(
      mapStateToProps
    )(FirebaseWrite)
  }
}
