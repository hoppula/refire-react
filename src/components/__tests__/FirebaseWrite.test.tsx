import * as React from "react"
import { Provider } from "react-redux"
import * as renderer from "react-test-renderer"
import { applyMiddleware, compose, createStore } from "redux"
import thunk from "redux-thunk"

import { firebaseActions } from "refire"
// actions are mocked in ../__mocks__/refire.ts
const { clearWriteErrors, write } = firebaseActions

import FirebaseWrite from "../FirebaseWrite"

function createEmptyStore() {
  return compose(
    applyMiddleware(thunk)
  )(createStore)((state, action) => {
    return {
      firebase: {
        writes: {
          errors: {
            "": ["Error"]
          },
          processing: {}
        }
      }
    }
  })
}

describe("FirebaseWrite", () => {
  it("invokes write action with submit", () => {

    interface CommentProps {
      submit?: (value: any) => void
    }
    function Comment({ submit }: CommentProps) {
      const submitComment = () => submit("Test")
      return (
        <button onClick={ submitComment } />
      )
    }

    const CommentComponent = FirebaseWrite({
      method: "update",
      path: ""
    })(Comment)

    const component = renderer.create(
      <Provider store={ createEmptyStore() }>
        <CommentComponent />
      </Provider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    tree.props.onClick({})
    expect(write.mock.calls.length).toBe(1)
  })

  it("clears write error with clearErrors", () => {

    interface ErrorsProps {
      errors?: string[],
      clearErrors?: () => void
    }
    function Errors({ errors, clearErrors }: ErrorsProps) {
      return (
        <button onClick={ clearErrors }>
          { errors[0] }
        </button>
      )
    }

    const ErrorsComponent = FirebaseWrite({
      method: "update",
      path: ""
    })(Errors)

    const component = renderer.create(
      <Provider store={ createEmptyStore() }>
        <ErrorsComponent />
      </Provider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    tree.props.onClick({})
    expect(clearWriteErrors.mock.calls.length).toBe(1)
  })

  it("errors are generated")
  it("processing prop gets updated")

})
