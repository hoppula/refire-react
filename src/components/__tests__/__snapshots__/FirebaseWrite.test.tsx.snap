// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirebaseWrite clears write error with clearErrors 1`] = `
<button
  onClick={[Function]}
>
  Error
</button>
`;

exports[`FirebaseWrite invokes write action with submit 1`] = `
<button
  onClick={[Function]}
/>
`;
